qiskit/transpiler/passes/synthesis/high_level_synthesis.py

```diff
@@ -0,0 +1,39 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+
+"""Synthesize high-level objects."""
+
+from qiskit.converters import circuit_to_dag
+from qiskit.transpiler.basepasses import TransformationPass
+from qiskit.dagcircuit.dagcircuit import DAGCircuit
+from qiskit.quantum_info.synthesis.clifford_decompose import decompose_clifford
+
+
+class HighLevelSynthesis(TransformationPass):
+    """Synthesize high-level objects by choosing the appropriate synthesis method based on
+    the object's name.
+    """
+
+    def run(self, dag: DAGCircuit) -> DAGCircuit:
+        """Run the HighLevelSynthesis pass on `dag`.
+        Args:
+            dag: input dag.
+        Returns:
+            Output dag with high level objects synthesized.
+        """
+
+        for node in dag.named_nodes("clifford"):
+            decomposition = circuit_to_dag(decompose_clifford(node.op))
+            dag.substitute_node_with_dag(node, decomposition)
+
+        return dag
```

Comment:
I would prefer we had more of the long-term structure set up in this pass before merge, but this is a less pressing concern to me than other parts of this PR, because it's easier for us to change transpiler-pass behaviour in subsequent releases.

Reply:
This pass seems to be a shim whose job is to call the synthesis routines (that live elsewhere) on nodes of interest. I think UnitarySynthesis is similar. And soon we will have synthesis of other higher-order operators. So I think we should consolidate them all under the same pass.